### ISSUE TITLE
First Try Feedback

### ISSUE THREAD

**Issue by defrex** [nhost/cli#6]
I just gave this a try and thought I'd collect my feedback here.

### How is it intended to be used?
You can only run `nhost init` on new directories. I assume most people will already have a project, however. 

It creates 3 things. It's not clear which of these should be in `.gitignore`, and which should be part of the project.

`docker-compose.example` - When I see a `.example` file, I expect that I should copy it (removing `.exmaple`) and add my customizations. The new file would then be gitignore'd. This expectation is broken here, since I'm supposed to alter `docker-compose.example` directly if I need customization, unless that customization is in `config.yaml`.

`config.yaml` - This file, presumably, should not be in source control (since it has secrets in it), thus it likely _does_ need a `.example` to allow other team members to create their own version (though of course I can easily create that). The name of the file, however, is a little awkward. I have config files for many things (next.js, Zeit now, eslint, prettier, etc.) in my repo, including a generic config for the env vars and secrets used by the project itself. Thus, `config.yaml` is a confusing name. Ideally something like `nhost.yaml` or `nhost.config.yaml` would be possible.

`migrations/` - 👍 

### `nhost dev`
I was not able to get this to work. However, I was (after some debugging) able to use the generated `docker-compose.yaml` file. Having this file be generated but exposed feels like a very leaky abstraction to me. Ideally, this is either an implementation detail (tucked inside a gitignore'd `.nhost` dir or something) or it's user-editable. Trying to do both is very unintuitive to me.

My preference would be for `nhost init` to generate a `docker-compose.yaml` file directly (scrap the `.example` & `config.yaml` files), but use env vars for secrets/configurability. Then the local user can simply put their local config into a `.env` file (a common practice for other things).

### Auth & Hasura
After I got my local setup working, the first thing I tried to do was import my configuration from Nhost using `hasura metadata export` and `hasura migrate create`. Both of those commands worked and generated the expected local migration files. However, when I tried to load them into my local instance I was unable. This is because the `auth` schema and associated tables are missing, and don't get pulled by hasura's export/migration functionality.

Ideally, `nhost init` would configure these tables as well.

Hopefully some of this feedback is useful. I absolutely love the idea of an easy-to-use local development experience for Nhost.

**Comment by nunopato** [comment]
@defrex thanks for the feedback, any of it at this point is very welcome! 

First of all, sorry for the lack of documentation about how to use it, the work here was done in a few hours and we didn't put too much thought into it since we mainly wanted to get a feel for how the development environment can look like.

**How is it intended to be used?** && **Nhost dev**

Running `nhost init` on existing directories is definitely something that we want to have. 

Regarding `docker-compose.example`, you are not supposed to edit it directly, you should edit `config.yaml` instead. I agree with you though that having `docker-compose.example` and `docker-compose.yaml (after nhost dev)` might be misleading and we might be better off omitting them completely. I myself had the same feeling. I will check whether docker compose allows for a string/stream instead of a file definition and if that doesn't play out, as you pointed out, we will abstract `docker-compose.yaml` (we don't even need to generate `docker-compose.example`) in a gitignore'd directory.

Also, I don't think `config.yaml` should be in source control (even though it can) because it is unique to your environment and automatically generated by `nhost init`. `graphql_admin_secret` is simply used to access your endpoint running locally (localhost) and actually, giving it a second thought, there is no real benefit on having it as a config parameter because `hasura-engine` is being launched with the console disabled. `config.yaml` is simply to give the user more control over where things are running and in what versions. `nhost.dev.yaml` is a way better name! Thanks for having me go over it again :)

The `migrations` folder should be pushed.

**Auth & Hasura**

That is, indeed, still missing. Coming soon! 

So, to summarize:
* We will allow for existing directories
* `nhost.dev.yaml` will simply hold configuration about versions and ports and the `admin_secret` will be removed because it is not needed.
* `nhost init` will only create `nhost.dev.yaml`, a `migrations` folder and a `.gitignore` with `nhost.dev.yaml` within it.

@defrex , does this make sense to you?

Also, where does your preference lean-to regarding `nhost dev` running detached from your console? would you prefer to have it attached like `now (zeit)`? and why? 

Thank you!

**Comment by defrex** [comment]
No worries about the fast first version. I'm a big fan of MVP strats, specifically because they elicit conversations like this one! 😄

Your proposed changes sound great to me. I do need access to `admin_secret`, however, since I require it for debug queries and for running a permissionless introspection query needed to power `graphql-codegen`. I don't need to specify it though, just know what it is.

Regarding detachment, my preference would be to keep it all in one proc. While I do typically run `docker-compose up` with `-d` for my other projects, you've proposed abstracting away the fact that you're using `docker-compose`. If this "feels" like a typical dev server, I'd rather have the simple mental model I'm used to for other dev servers. That is to say: `nhost dev` is simpler to manage than `nhost dev up`, `nhost dev ps`, `nhost dev stop`, and `nhost dev down` which would all be needed to map `docker-compose`. It also feels likely to end up a leaky abstraction that way.

Looking forward to seeing more! I eagerly await auth support so I can start using it for my own local development.

**Comment by nunopato** [comment]
@defrex indeed, these are the most fruitful 😄

Would an optional `graphql_admin_secret` config parameter work? if specified, `nhost dev` uses it, otherwise, it generates a new one on the fly on every `nhost dev` outputting it to stdout? 

Coming back to `config.yaml`, it just came back to me again why, unfortunately, we can't use another filename - it is an actual requirement for `hasura console`. I will suggest to hasura to make this optional since the only required config parameter is `endpoint` and it can be passed in as a command flag: `hasura console --endpoint=http://localhost:8080`. Until then we have to live with it. 

Attached vs detached - I like the reasoning 👍 

Thank you!

**Comment by defrex** [comment]
An optional parameter for `graphql_admin_secret` would totally work. 👍 

That's a shame about `hasura-cli`, but it seems there isn't much to be done about it.

**Comment by nunopato** [comment]
@defrex We just published a new version of the CLI that incorporates some of the changes we discussed as well the auth schema out-of-the-box. As always, any feedback is much appreciated.

https://www.npmjs.com/package/nhost

**Comment by defrex** [comment]
Awesome, this version is much improved. Great work.

One thing worth pointing out: if the services fail to boot, the `nhost dev` command gets stuck in this loop indefinitely. Ctrl+C wouldn't even kill it, I had to `kill` it. A timeout + not locking the proc would be a nice improvement.
https://github.com/nhost/cli/blob/master/src/commands/dev.js#L111 

Once I figured that out (via `cd .nhost && docker-componse logs`), I was able to solve the issue and get `nhost dev` working as expected. Which is a solid step forward.

However, my next move was to import my configuration from Nhost in order to begin using the local server with my client. Here is a replay of that experience.

```bash
╔══ defrex@Arons-MacBook-Pro.local ~/Code/servehub:master!11
╚╡hasura metadata export --endpoint=https://hasura-[redacted].nhost.app --admin-secret=[redacted]
0da8
INFO Metadata exported                            
╔══ defrex@Arons-MacBook-Pro.local ~/Code/servehub:master!11
╚╡hasura migrate create  "update" --from-server --endpoint=https://hasura-[redacted].nhost.app --admin-secret=[redacted]
INFO Migrations files created                      name=update version=1586455398157
╔══ defrex@Arons-MacBook-Pro.local ~/Code/servehub:master!12
╚╡hasura migrate apply --endpoint=http://localhost:8080 --admin-secret=[redacted]
[0000] apply failed: [postgres-error] query execution failed ($.args[0].args)
File: '1586455398157_update/up.sql'
{
    "sql": "CREATE FUNCTION public.set_current_timestamp_updated_at() RETURNS trigger\n    LANGUAGE plpgsql\n    AS $$\nDECLARE\n  _new record;\nBEGIN\n  _new := NEW;\n  _new.\"updated_at\" = NOW();\n  RETURN _new;\nEND;\n$$;\nCREATE TABLE public.attempts (\n    id uuid DEFAULT public.gen_random_uuid() NOT NULL,\n    created_at timestamp with time zone DEFAULT now() NOT NULL,\n    updated_at timestamp with time zone DEFAULT now() NOT NULL,\n    attempted_at timestamp with time zone NOT NULL,\n    image_url text,\n    job_id uuid NOT NULL,\n    success boolean NOT NULL\n);\nCREATE TABLE public.documents (\n    id uuid DEFAULT public.gen_random_uuid() NOT NULL,\n    created_at timestamp with time zone DEFAULT now() NOT NULL,\n    updated_at timestamp with time zone DEFAULT now() NOT NULL,\n    pickup boolean DEFAULT true NOT NULL,\n    street text,\n    unit text,\n    postal_code text,\n    city text,\n    province text,\n    job_id uuid NOT NULL,\n    title text NOT NULL,\n    url text\n);\nCREATE TABLE public.jobs (\n    id uuid DEFAULT public.gen_random_uuid() NOT NULL,\n    created_at timestamp with time zone DEFAULT now() NOT NULL,\n    updated_at timestamp with time zone DEFAULT now() NOT NULL,\n    lawyer_user_id uuid NOT NULL,\n    server_user_id uuid,\n    description text,\n    target_id uuid,\n    stripe_payment_intent_id text,\n    stripe_payment_intent_client_secret text,\n    stripe_payment_intent_succeeded boolean DEFAULT false NOT NULL\n);\nCREATE TABLE public.messages (\n    id uuid DEFAULT public.gen_random_uuid() NOT NULL,\n    created_at timestamp with time zone DEFAULT now() NOT NULL,\n    updated_at timestamp with time zone DEFAULT now() NOT NULL,\n    message text NOT NULL,\n    user_id uuid NOT NULL,\n    job_id uuid NOT NULL\n);\nCREATE TABLE public.roles (\n    role text NOT NULL\n);\nCREATE TABLE public.targets (\n    id uuid DEFAULT public.gen_random_uuid() NOT NULL,\n    created_at timestamp with time zone DEFAULT now() NOT NULL,\n    updated_at timestamp with time zone DEFAULT now() NOT NULL,\n    name text NOT NULL,\n    street text NOT NULL,\n    unit text NOT NULL,\n    city text NOT NULL,\n    province character varying NOT NULL,\n    postal_code text NOT NULL\n);\nCREATE TABLE public.user_roles (\n    id uuid DEFAULT public.gen_random_uuid() NOT NULL,\n    created_at timestamp with time zone DEFAULT now() NOT NULL,\n    user_id uuid NOT NULL,\n    role text NOT NULL\n);\nCREATE TABLE public.users (\n    id uuid DEFAULT public.gen_random_uuid() NOT NULL,\n    created_at timestamp with time zone DEFAULT now() NOT NULL,\n    updated_at timestamp with time zone DEFAULT now() NOT NULL,\n    display_name text NOT NULL,\n    active boolean DEFAULT false NOT NULL,\n    default_role text DEFAULT 'user'::text NOT NULL,\n    avatar_url text,\n    email text,\n    secret_token uuid DEFAULT public.gen_random_uuid() NOT NULL,\n    secret_token_expires_at timestamp with time zone DEFAULT now() NOT NULL,\n    register_data jsonb,\n    is_anonymous boolean DEFAULT false NOT NULL,\n    stripe_customer_id text,\n    firebase_messaging_token text,\n    name text\n);\nALTER TABLE ONLY public.attempts\n    ADD CONSTRAINT attempts_pkey PRIMARY KEY (id);\nALTER TABLE ONLY public.documents\n    ADD CONSTRAINT documents_pkey PRIMARY KEY (id);\nALTER TABLE ONLY public.jobs\n    ADD CONSTRAINT jobs_pkey PRIMARY KEY (id);\nALTER TABLE ONLY public.jobs\n    ADD CONSTRAINT jobs_target_id_key UNIQUE (target_id);\nALTER TABLE ONLY public.messages\n    ADD CONSTRAINT messages_pkey PRIMARY KEY (id);\nALTER TABLE ONLY public.roles\n    ADD CONSTRAINT roles_pkey PRIMARY KEY (role);\nALTER TABLE ONLY public.targets\n    ADD CONSTRAINT target_pkey PRIMARY KEY (id);\nALTER TABLE ONLY public.targets\n    ADD CONSTRAINT targets_id_key UNIQUE (id);\nALTER TABLE ONLY public.user_roles\n    ADD CONSTRAINT user_roles_pkey PRIMARY KEY (id);\nALTER TABLE ONLY public.user_roles\n    ADD CONSTRAINT user_roles_user_id_role_key UNIQUE (user_id, role);\nALTER TABLE ONLY public.users\n    ADD CONSTRAINT users_email_key UNIQUE (email);\nALTER TABLE ONLY public.users\n    ADD CONSTRAINT users_pkey PRIMARY KEY (id);\nALTER TABLE ONLY public.users\n    ADD CONSTRAINT users_secret_token_key UNIQUE (secret_token);\nCREATE TRIGGER set_public_attempts_updated_at BEFORE UPDATE ON public.attempts FOR EACH ROW EXECUTE FUNCTION public.set_current_timestamp_updated_at();\nCOMMENT ON TRIGGER set_public_attempts_updated_at ON public.attempts IS 'trigger to set value of column \"updated_at\" to current timestamp on row update';\nCREATE TRIGGER set_public_documents_updated_at BEFORE UPDATE ON public.documents FOR EACH ROW EXECUTE FUNCTION public.set_current_timestamp_updated_at();\nCOMMENT ON TRIGGER set_public_documents_updated_at ON public.documents IS 'trigger to set value of column \"updated_at\" to current timestamp on row update';\nCREATE TRIGGER set_public_jobs_updated_at BEFORE UPDATE ON public.jobs FOR EACH ROW EXECUTE FUNCTION public.set_current_timestamp_updated_at();\nCOMMENT ON TRIGGER set_public_jobs_updated_at ON public.jobs IS 'trigger to set value of column \"updated_at\" to current timestamp on row update';\nCREATE TRIGGER set_public_messages_updated_at BEFORE UPDATE ON public.messages FOR EACH ROW EXECUTE FUNCTION public.set_current_timestamp_updated_at();\nCOMMENT ON TRIGGER set_public_messages_updated_at ON public.messages IS 'trigger to set value of column \"updated_at\" to current timestamp on row update';\nCREATE TRIGGER set_public_target_updated_at BEFORE UPDATE ON public.targets FOR EACH ROW EXECUTE FUNCTION public.set_current_timestamp_updated_at();\nCOMMENT ON TRIGGER set_public_target_updated_at ON public.targets IS 'trigger to set value of column \"updated_at\" to current timestamp on row update';\nCREATE TRIGGER set_public_users_updated_at BEFORE UPDATE ON public.users FOR EACH ROW EXECUTE FUNCTION public.set_current_timestamp_updated_at();\nALTER TABLE ONLY public.attempts\n    ADD CONSTRAINT attempts_job_id_fkey FOREIGN KEY (job_id) REFERENCES public.jobs(id) ON UPDATE CASCADE ON DELETE CASCADE;\nALTER TABLE ONLY public.documents\n    ADD CONSTRAINT documents_job_id_fkey FOREIGN KEY (job_id) REFERENCES public.jobs(id) ON UPDATE CASCADE ON DELETE CASCADE;\nALTER TABLE ONLY public.jobs\n    ADD CONSTRAINT jobs_lawyer_user_id_fkey FOREIGN KEY (lawyer_user_id) REFERENCES public.users(id) ON UPDATE CASCADE ON DELETE CASCADE;\nALTER TABLE ONLY public.jobs\n    ADD CONSTRAINT jobs_server_user_id_fkey FOREIGN KEY (server_user_id) REFERENCES public.users(id) ON UPDATE CASCADE ON DELETE CASCADE;\nALTER TABLE ONLY public.jobs\n    ADD CONSTRAINT jobs_target_id_fkey FOREIGN KEY (target_id) REFERENCES public.targets(id) ON UPDATE CASCADE ON DELETE RESTRICT;\nALTER TABLE ONLY public.messages\n    ADD CONSTRAINT messages_job_id_fkey FOREIGN KEY (job_id) REFERENCES public.jobs(id) ON UPDATE CASCADE ON DELETE CASCADE;\nALTER TABLE ONLY public.messages\n    ADD CONSTRAINT messages_user_id_fkey FOREIGN KEY (user_id) REFERENCES public.users(id) ON UPDATE CASCADE ON DELETE CASCADE;\nALTER TABLE ONLY public.user_roles\n    ADD CONSTRAINT user_roles_role_fkey FOREIGN KEY (role) REFERENCES public.roles(role) ON UPDATE RESTRICT ON DELETE CASCADE;\nALTER TABLE ONLY public.user_roles\n    ADD CONSTRAINT user_roles_user_id_fkey FOREIGN KEY (user_id) REFERENCES public.users(id) ON UPDATE RESTRICT ON DELETE CASCADE;\nALTER TABLE ONLY public.users\n    ADD CONSTRAINT users_default_role_fkey FOREIGN KEY (default_role) REFERENCES public.roles(role) ON UPDATE RESTRICT ON DELETE RESTRICT;\n"
}
[42723] FatalError: function "set_current_timestamp_updated_at" already exists with same argument types
```

The issue occurs because the imported migration wants to create things already created by the built-in migration. If it's possible to separate the `auth` schema migration from the `public` schema migration during the `init` phase, I would be able to replace the second migration with my imported one to solve the issue. There may be other approaches though, I'm not sure if Hasura has built-in accommodation for importing a migration delta.

**Comment by nunopato** [comment]
@defrex thanks again!

yes, the current implementation doesn't assume things going south, yet :) what was the issue in the end?

We will work next on having `nhost init` behave correctly when there is already a setup for migrations, because right now the assumption is that there is not `migrations` folder there.

Thanks

**Comment by defrex** [comment]
The issue was the same as the one I posted, more or less. I already had a migrations directory with an init migration.

Looking forward to the next release!

**Comment by elitan** [comment]
Very good feedback. Me and @nunopato talked about exactly this today. How to get `nhost` to work with already existing projects.

Our idea here was to add `endpoint` and `admin-secret` to `nhost init` like this:

`nhost init --endpoint=https://hasura-[redacted].nhost.app --admin-secret=[reducted]`

This enables us to download and use all existing migrations from an existing project, instead of using the default migrations on a regular `nhost init`. Internally we generate a `config.yaml` file and execute the following command: `hasura migrate create "init" --from-server`.

Reference: https://hasura.io/docs/1.0/graphql/manual/migrations/existing-database.html#step-3-initialize-the-migrations-as-per-your-current-state

**Comment by defrex** [comment]
That sounds like exactly what I need. Great progress guys. 💪

**Comment by nunopato** [comment]
@defrex now looking again at the commands you issued, I am not sure I understand what you were trying to achieve. 

1) `hasura metadata export --endpoint=https://hasura-[redacted].nhost.app --admin-secret=[redacted]` - this exports your current metadata

2) `hasura migrate create  "update" --from-server --endpoint=https://hasura-[redacted].nhost.app --admin-secret=[redacted]` - this initializes or creates migrations as per your current state

3) `hasura migrate apply --endpoint=http://localhost:8080 --admin-secret=[redacted]` - this tries to apply the migration file you just created from 2). 

If I understood all steps, the error you got makes all sense. What am I missing here?

I think what happened when your console froze, is that you already had a non-empty schema with your Postgres installation and the migration file we ship with the CLI (`{timestamp}_init`) is a snapshot migration of our most up-to-date auth/public schemas which is well suited for a clean-slate project/schema but not if you already have a local schema and only an incremental (or delta) approach would work. 

@elitan the link you posted (migrations-existing-db) is what I am using to generate the `init` migration that we ship and can be used to generate the migration locally... but the more difficult problem is when things get out of sync because of different versions/schemas.

**Comment by elitan** [comment]
> 2. A user has already an existing schema. As mentioned, we can create a migration file based on that current state but if the user changes the version of HBP currently in use, that might introduce inconsistencies between the code and the schema.

Above all, we should really try to **not** change the schema for HBP so we should really keep this problem to a minimum from the start.

Hopefully, we could provide raw SQL scripts to upgrade an HBP version. These scripts can be run locally for a user on their dev env and produce new migrations entries.

![image](https://user-images.githubusercontent.com/331818/78963424-c216b400-7af7-11ea-8042-c4f8f03c952b.png)

**Comment by defrex** [comment]
@nunopato 
> If I understood all steps, the error you got makes sense. What am I missing here?

Nothing! 😅 The point of posting the log was to explain what I'd like to do and demonstrate why it doesn't work. You picked up what I was putting down.

**Comment by nunopato** [comment]
@defrex thank you! will close the issue now.
